test/src/tests-batchUploader.ts

```diff
@@ -84,16 +84,603 @@ describe('batch uploader', () => {
+                    'https://jssdks.mparticle.com/v3/JS/test_key/events',
+                    400
+                );
+
+                const newLogger = new Logger(window.mParticle.config);
+                const mpInstance = window.mParticle.getInstance();
+
+                const uploader = new BatchUploader(mpInstance, 1000);
+
+                const batchValidator = new _BatchValidator();
+
+                const batch1 = batchValidator.returnBatch({
+                    messageType: 4,
+                    name: 'Test Event 1',
+                });
+
+                const batch2 = batchValidator.returnBatch({
+                    messageType: 4,
+                    name: 'Test Event 2',
+                });
+                const batch3 = batchValidator.returnBatch({
+                    messageType: 4,
+                    name: 'Test Event 3',
+                });
+
+                // HACK: Directly access uploader to Force an upload
+                const batchesNotUploaded = await (<any>uploader).uploadBatches(
+                    newLogger,
+                    [batch1, batch2, batch3],
+                    false
+                );
+
+                expect(batchesNotUploaded).to.be.ok;
+
+                expect(
+                    batchesNotUploaded.length,
+                    'Should have 3 uploaded batches'
+                ).to.equal(3);
+
+                expect(
+                    batchesNotUploaded[0].events[0].data.event_name
+                ).to.equal('Test Event 1');
+                expect(
+                    batchesNotUploaded[1].events[0].data.event_name
+                ).to.equal('Test Event 2');
+                expect(
+                    batchesNotUploaded[2].events[0].data.event_name
+                ).to.equal('Test Event 3');
+            });
+        });
+    });
+
+    describe('Upload Workflow', () => {
+        beforeEach(() => {
+            clock = sinon.useFakeTimers({
+                now: new Date().getTime(),
+            });
+        });
+
+        afterEach(() => {
+            window.fetchMock.restore();
+            clock.restore();
+        });
+
+        it('should organize events and batches in the order they are processed and maintain that order when uploading', (done) => {
+            // Batches should be uploaded in the order they were created to prevent
+            // any potential corruption.
+
+            window.fetchMock.post(urls.eventsV3, 200);
+            window.fetchMock.config.overwriteRoutes = true;
+
+            window.mParticle.config.flags = {
+                eventsV3: '100',
+                eventBatchingIntervalMillis: 1000,
+            };
+
+            window.mParticle._resetForTests(MPConfig);
+            window.mParticle.init(apiKey, window.mParticle.config);
+
+            window.mParticle.logEvent('Test Event 0');
+
+            // Force time to pass so that batch processing occurs
+            clock.tick(1000);
```

Comment:
I would stick to either doing `mParticle.upload` or `clock.tick` throughout all the tests to avoid confusion as to why you are doing one and not the other.  Do a full find/replace in this file.